File: modules/pkgs/registration/registration.go
// SPDX-FileCopyrightText: 2024-2026 TII (SSRC) and the Ghaf contributors
// SPDX-License-Identifier: Apache-2.0

// Service registration functionality for the GIVC agent.
package registration

import (
	"context"
	"fmt"
	"strings"
	"sync"
	"time"

	givc_admin "givc/modules/api/admin"
	givc_systemd "givc/modules/api/systemd"
	givc_config "givc/modules/pkgs/config"
	givc_serviceclient "givc/modules/pkgs/serviceclient"
	givc_servicemanager "givc/modules/pkgs/servicemanager"

	log "github.com/sirupsen/logrus"
)

// RegistrationConfig holds the configuration needed for service registration
type RegistrationConfig struct {
	SystemdServer *givc_servicemanager.SystemdControlServer
	AgentConfig   *givc_config.AgentConfig
}

// Registry defines the interface for service registration operations
type Registry interface {
	// RegisterAgent registers the agent with the admin server
	RegisterAgent(ctx context.Context) error

	// RegisterServices registers all configured services with the admin server
	RegisterServices(ctx context.Context) error

	// StartRegistrationWorker starts the registration process
	StartRegistrationWorker(ctx context.Context, wg *sync.WaitGroup, serverStarted <-chan struct{})
}

// ServiceRegistry implements the Registry interface
type ServiceRegistry struct {
	config RegistrationConfig
}

// NewServiceRegistry creates a new ServiceRegistry with the given configuration
func NewServiceRegistry(config RegistrationConfig) Registry {
	return &ServiceRegistry{
		config: config,
	}
}

// StartRegistrationWorker starts the registration process
func (r *ServiceRegistry) StartRegistrationWorker(ctx context.Context, wg *sync.WaitGroup, serverStarted <-chan struct{}) {
	wg.Add(1)
	go func() {
		defer wg.Done()

		// Wait for server to start to handle callbacks
		select {
		case <-serverStarted:
		case <-ctx.Done():
			log.Infof("Registration cancelled before server start")
			return
		}

		// Register agent
		if err := r.RegisterAgent(ctx); err != nil {
			log.Errorf("Failed to register agent: %v", err)
			return
		}

		// Register services (systemd units)
		if err := r.RegisterServices(ctx); err != nil {
			log.Errorf("Failed to register services: %v", err)
			return
		}
		log.Infof("Registration goroutine finished")
	}()
}

// RegisterAgent registers the agent with the admin server
func (r *ServiceRegistry) RegisterAgent(ctx context.Context) error {
	if r.config.SystemdServer == nil {
		return fmt.Errorf("systemd server not configured")
	}

	agentServiceName := r.config.AgentConfig.Identity.ServiceName
	if agentServiceName == "" {
		return fmt.Errorf("agent service name not configured")
	}

	unitStatus, err := r.config.SystemdServer.GetUnitStatus(ctx, &givc_systemd.UnitRequest{
		UnitName: agentServiceName,
	})
	if err != nil {
		return err
	}

	agentEntryRequest := &givc_admin.RegistryRequest{
		Name:   agentServiceName,
		Type:   r.config.AgentConfig.Identity.Type,
		Parent: r.config.AgentConfig.Identity.Parent,
		Transport: &givc_admin.TransportConfig{
			Protocol: r.config.AgentConfig.Network.AgentEndpoint.Transport.Protocol,
			Address:  r.config.AgentConfig.Network.AgentEndpoint.Transport.Address,
			Port:     r.config.AgentConfig.Network.AgentEndpoint.Transport.Port,
			Name:     r.config.AgentConfig.Network.AgentEndpoint.Transport.Name,
		},
		State: unitStatus.UnitStatus,
	}

	// Register agent with admin server with retry loop
	return r.registerWithRetry(ctx, agentEntryRequest, "agent")
}

// RegisterServices registers all configured services with the admin server
func (r *ServiceRegistry) RegisterServices(ctx context.Context) error {
	for service, subType := range r.config.AgentConfig.Capabilities.Units {
		if !strings.HasSuffix(service, ".service") {
			continue
		}

		select {
		case <-ctx.Done():
			log.Infof("Service registration cancelled")
			return ctx.Err()
		default:
		}

		if err := r.registerSingleService(ctx, service, subType); err != nil {
			log.Warnf("Failed to register service %s: %v", service, err)
			// Continue with other services even if one fails
			continue
		}
	}

	return nil
}

// registerSingleService registers a single service with the admin server
func (r *ServiceRegistry) registerSingleService(ctx context.Context, service string, subType uint32) error {
	if r.config.SystemdServer == nil {
		return fmt.Errorf("cannot register service %s: systemd server not configured", service)
	}

	unitStatus, err := r.config.SystemdServer.GetUnitStatus(ctx, &givc_systemd.UnitRequest{
		UnitName: service,
	})
	if err != nil {
		log.Warnf("Error getting unit status for %s: %s", service, err)
		return err
	}

	serviceEntryRequest := &givc_admin.RegistryRequest{
		Name:   service,
		Parent: r.config.AgentConfig.Identity.ServiceName,
		Type:   uint32(subType),
		Transport: &givc_admin.TransportConfig{
			Name:     r.config.AgentConfig.Network.AgentEndpoint.Transport.Name,
			Protocol: r.config.AgentConfig.Network.AgentEndpoint.Transport.Protocol,
			Address:  r.config.AgentConfig.Network.AgentEndpoint.Transport.Address,
			Port:     r.config.AgentConfig.Network.AgentEndpoint.Transport.Port,
		},
		State: unitStatus.UnitStatus,
	}

	log.Infof("Trying to register service: %s", service)
	_, err = givc_serviceclient.RegisterRemoteService(r.config.AgentConfig.Network.AdminEndpoint, serviceEntryRequest)
	if err != nil {
		log.Warnf("Error registering service %s: %s", service, err)
		return err
	} else {
		log.Infof("Successfully registered service: %s", service)
	}

	return nil
}

// registerWithRetry performs registration with retry logic
func (r *ServiceRegistry) registerWithRetry(ctx context.Context, request *givc_admin.RegistryRequest, entityType string) error {
	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
			_, err := givc_serviceclient.RegisterRemoteService(r.config.AgentConfig.Network.AdminEndpoint, request)
			if err == nil {
				log.Infof("Successfully registered %s: %s", entityType, request.Name)
				return nil
			}
			log.Warnf("Error registering %s: %s, retrying...", entityType, err)
			time.Sleep(1 * time.Second)
		}
	}
}

File: crates/agent/src/registration.rs
// SPDX-FileCopyrightText: 2025-2026 TII (SSRC) and the Ghaf contributors
// SPDX-License-Identifier: Apache-2.0

use std::time::Duration;

use anyhow::{Result, bail};
use givc_client::AdminClient;
use givc_common::address::EndpointAddress;
use givc_common::types::{EndpointEntry, UnitStatus};
use tokio::sync::oneshot;
use tokio::time::sleep;
use tracing::{info, warn};

use crate::config::{AgentConfig, EndpointConfig as AgentEndpointConfig};
use crate::servicemanager::{Snapshot, SystemdBackend, ZbusBackend};

pub fn start_registration_worker(
    config: AgentConfig,
    backend: ZbusBackend,
    server_started: oneshot::Receiver<()>,
) {
    tokio::spawn(async move {
        if server_started.await.is_err() {
            return;
        }

        if let Err(err) = register_agent_with_retry(&config, &backend).await {
            warn!(error = %err, "failed to register agent");
            return;
        }

        if let Err(err) = register_services(&config, &backend).await {
            warn!(error = %err, "failed to register services");
            return;
        }

        info!("registration goroutine finished");
    });
}

async fn register_agent_with_retry(config: &AgentConfig, backend: &ZbusBackend) -> Result<()> {
    loop {
        match register_agent(config, backend).await {
            Ok(()) => return Ok(()),
            Err(err) => {
                warn!(error = %err, "error registering agent, retrying...");
                sleep(Duration::from_secs(1)).await;
            }
        }
    }
}

async fn register_agent(config: &AgentConfig, backend: &ZbusBackend) -> Result<()> {
    let agent_service_name = config.identity.service_name.clone();
    if agent_service_name.is_empty() {
        bail!("agent service name not configured");
    }

    let admin = admin_client(config)?;
    let unit_status = backend.get_unit_snapshot(&agent_service_name).await?;
    admin
        .register_service(
            agent_service_name,
            config.identity.r#type.try_into()?,
            endpoint_entry(&config.network.agent)?,
            snapshot_to_unit_status(unit_status),
        )
        .await?;
    info!("successfully registered agent");
    Ok(())
}

async fn register_services(config: &AgentConfig, backend: &ZbusBackend) -> Result<()> {
    let admin = admin_client(config)?;
    for (service, sub_type) in &config.capabilities.units {
        if !service.ends_with(".service") {
            continue;
        }

        match backend.get_unit_snapshot(service).await {
            Ok(snapshot) => {
                if let Err(err) = admin
                    .register_service(
                        service.clone(),
                        (*sub_type).try_into()?,
                        endpoint_entry(&config.network.agent)?,
                        snapshot_to_unit_status(snapshot),
                    )
                    .await
                {
                    warn!(service = %service, error = %err, "error registering service");
                    continue;
                }
                info!(service = %service, "successfully registered service");
            }
            Err(err) => {
                warn!(service = %service, error = %err, "error getting unit status");
            }
        }
    }

    Ok(())
}

fn admin_client(config: &AgentConfig) -> Result<AdminClient> {
    let admin_tls_name = if config.network.admin.transport.name.is_empty() {
        "admin.ghaf".to_owned()
    } else {
        config.network.admin.transport.name.clone()
    };
    let admin_tls = config
        .network
        .tls_config
        .clone()
        .map(|tls| (admin_tls_name, tls));
    Ok(AdminClient::from_endpoint_address(
        endpoint_address(&config.network.admin.transport)?,
        admin_tls,
    ))
}

fn endpoint_address(endpoint: &crate::config::TransportConfig) -> Result<EndpointAddress> {
    Ok(match endpoint.protocol.as_str() {
        "tcp" => EndpointAddress::Tcp {
            addr: endpoint.address.clone(),
            port: endpoint.port.parse()?,
        },
        "unix" => EndpointAddress::Unix(endpoint.address.clone()),
        "abstract" => EndpointAddress::Abstract(endpoint.address.clone()),
        "vsock" => EndpointAddress::Vsock(tokio_vsock::VsockAddr::new(
            endpoint.address.parse()?,
            endpoint.port.parse()?,
        )),
        other => bail!("unsupported admin transport protocol: {other}"),
    })
}

fn endpoint_entry(endpoint: &AgentEndpointConfig) -> Result<EndpointEntry> {
    Ok(EndpointEntry {
        address: endpoint_address(&endpoint.transport)?,
        tls_name: endpoint.transport.name.clone(),
    })
}

fn snapshot_to_unit_status(snapshot: Snapshot) -> UnitStatus {
    UnitStatus {
        name: snapshot.name,
        description: snapshot.description,
        load_state: snapshot.load_state,
        active_state: snapshot.active_state,
        sub_state: snapshot.sub_state,
        path: snapshot.path,
        freezer_state: snapshot.freezer_state,
    }
}
